### PR TITLE
add custom DNS Resolver configuration

### PR DESCRIPTION
**Pull request by vyzo**

### CODE REVIEW

File: dns.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package config
+
+// DNSConfig specifies custom resolvers using DoH
+type DNSConfig struct {
+	// DefaultResolver, if present, is a URL for the default DoH resolver.
+	// If empty, DNS resolution will use the system resolver.
+	DefaultResolver string `json:",omitempty"`
+	// CustomResolvers is a map of domains to URLs for custom DoH resolution.
+	CustomResolvers map[string]string `json:",omitempty"`
```
Comment (Stebalien) [inline dns.go:9]: Do we _need_ a default resolver? I'd assume that we could have a single "." entry in a `Resolvers` map, right?

Also, let's briefly (very briefly) document the format (i.e., URL? Multiaddr?).

**Comment by vyzo** [comment]
In principle we could, but it seems more user friendly to have an explicit
default.

**Comment by vyzo** [comment]
and sure, i'll add some documentation.

**Comment by Stebalien** [comment]
Both SGTM.

File: dns.go (hunk 1 above)
Comment (aschmahmann) [reply dns.go:9]: note: some documentation in code can happen here, but the user facing docs live in go-ipfs https://github.com/ipfs/go-ipfs/blob/master/docs/config.md

Comment (lidel) [reply dns.go:9]: +1  on removing `DefaultResolver` and renaming `CustomResolvers` to  `FQDNResolvers` to be extra clear about what the "custom"  rules are about. (I'd like  to have [FQDN](https://en.wikipedia.org/wiki/Fully_qualified_domain_name) in name so people familiar with DNS will immediately understand format and convention for keys).

## Don't skip in JSON

We should avoid `json:",omitempty"` here. 
It is better for discoverability to have empty `DNS.FQDNResolvers` (ok to have `{}`) – that is what we already do in `Pinning.RemoteServices`. 


### Values 

We could support multiaddrs and URLs, but for DoH  URL will be what people want to copy&paste while setting it up, so ok with saying this needs to be https:// URL with DoH  endpoint.

### Keys

I was thinking about interop with wider DNS ecosystem here and suggest we don't invent any new notation with `*` as I initially syggested, but follow the existing DNS convention of using the [explicit FQDN syntax](https://en.wikipedia.org/wiki/Fully_qualified_domain_name#Syntax) in our config files.

To illustrate:

```json
  "DNS": {
    "FQDNResolvers": {
      "eth.": "https://eth.link-or-cloudflare/path/to/dns-query",
      "crypto.": "https://unstoppablesomething.example.com/path/to/dns-query",
      "libre.": "https://www.opennic.org/path/to/their/dns-query", 
      ".": "https://mozilla.cloudflare-dns.com/dns-query"
    }
  }
```

Idea here is to closely follow the format of [zone files](https://en.wikipedia.org/wiki/Zone_file), which should make things easier to understand.  (The DNS root is unnamed, expressed as the empty label terminated by the dot.)


We should enforce keys are explicit FQDNs by validating them with [`dns.IsFqdn`](https://github.com/miekg/dns/blob/83b388a80c430233c72c45c85517573177b7f716/defaults.go#L269) – removes surface for DNS ambiguity.

**Review by lidel — CHANGES_REQUESTED** [review]
Some asks  in comment inline. 

Rationale: I'd like to avoid inventing custom syntax in keys and concepts, if we can.
DNS ecosystem and tooling already has pre-existing concepts like FQDN and configuration in form of Zone files, we should follow what exists where possible to improve onboarding.

File: dns.go (hunk 1 above)
Comment (vyzo) [reply dns.go:9]: ok fair enough; we can call the field just `Resolvers` however.

Comment (vyzo) [reply dns.go:9]: I will also make some minor changes in the madns Resolver to normalize to fqdns.

File: dns.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package config
+
+// DNSConfig specifies DNS resolution rules using custom resolvers
+type DNSConfig struct {
+	// CustomResolvers is a map of FQDNs to URLs for custom DNS resolution.
```
Comment (lidel) [inline dns.go:5]:
```suggestion
	// Resolvers is a map of FQDNs to URLs for custom DNS resolution. 
	// URLs starting with `https://` indicate DoH endpoints.
	// Support for other resolver types can be added in the future.
	// https://en.wikipedia.org/wiki/Fully_qualified_domain_name
	// https://en.wikipedia.org/wiki/DNS_over_HTTPS 
	// 
	// Example:
	// - Custom resolver for ENS:          `eth.` → `https://eth.link/dns-query`
	// - Override the default OS resolver: `.`    → `https://doh.applied-privacy.net/query`
```

**Review by lidel — APPROVED** [review]
LGTM, but please update the comment.

**Review by Stebalien — APPROVED** [review]

**Pull request merged**